# 1. Data upload and verification

Import of all necessary libraries:

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.metrics import roc_auc_score
from sklearn.tree import DecisionTreeClassifier, export_graphviz
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

## 1.1. Training purpose data

Data upload for model training purpose with cell execution time determination:

In [3]:
%%time
train_df = pd.read_csv('valami_train.txt.gz', sep='\t', header = None)

Wall time: 1min 32s


Taking a look what's inside:

In [3]:
train_df.head()

0       1    2    3    4    5    6    7    8    9     ...       729  \
0    0  0.4286    0    1    0    0    0    0    0    0   ...    0.8086   
1    0  0.7143    0    1    0    0    0    0    0    0   ...    0.9570   
2    0  0.5000    0    1    0    0    0    0    0    0   ...    0.8945   
3    0  0.6429    0    0    0    1    0    0    0    0   ...    0.9570   
4    0  0.3571    0    1    0    0    0    0    0    0   ...    0.9492   

      730     731     732  733     734  735     736  737     738  
0  0.0020  0.3478  0.8450    0  0.7045    0  0.0101    0  0.0123  
1  0.0020  0.5217  0.9441    0  0.8832    0  0.0074    0  0.0075  
2  0.0020  0.2174  0.8883    0  0.7560    0  0.0018    0  0.0101  
3  0.0011  0.5217  0.9243    0  0.8419    0  0.0064    0  0.0049  
4  0.0025  0.4348  0.8450    0  0.7045    0  0.0101    0  0.0144  

[5 rows x 739 columns]

Obtaining data frame size:

In [20]:
train_df.shape

(190000, 739)

Classifier data - simple binary values:

In [36]:
train_df[0].unique()

array([0, 1], dtype=int64)

Data verification for columns using DataFrame information function:
data in all columns that we have are numeric

In [13]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190000 entries, 0 to 189999
Columns: 739 entries, 0 to 738
dtypes: float64(325), int64(414)
memory usage: 1.0 GB


Some general information about columns values (maybe will help us later):

In [38]:
train_df.describe(include='all')

0              1              2              3    \
count  190000.00000  190000.000000  190000.000000  190000.000000   
mean        0.05080       0.600336       0.428968       0.269347   
std         0.21959       0.245293       0.494930       0.443622   
min         0.00000       0.029800       0.000000       0.000000   
25%         0.00000       0.357100       0.000000       0.000000   
50%         0.00000       0.642900       0.000000       0.000000   
75%         0.00000       0.857100       1.000000       1.000000   
max         1.00000       1.000000       1.000000       1.000000   

                 4              5              6              7    \
count  190000.000000  190000.000000  190000.000000  190000.000000   
mean        0.177968       0.086611       0.008789       0.019516   
std         0.382487       0.281264       0.093340       0.138329   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                 8              9        ...                  729  \
count  190000.000000  190000.000000      ...        190000.000000   
mean        0.000158       0.003811      ...             0.895043   
std         0.012565       0.061612      ...             0.079869   
min         0.000000       0.000000      ...             0.000000   
25%         0.000000       0.000000      ...             0.886700   
50%         0.000000       0.000000      ...             0.921900   
75%         0.000000       0.000000      ...             0.945300   
max         1.000000       1.000000      ...             1.000000   

                 730            731            732            733  \
count  190000.000000  190000.000000  190000.000000  190000.000000   
mean        0.001731       0.375561       0.903889       0.000021   
std         0.001395       0.171971       0.067851       0.004588   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.001000       0.260900       0.839600       0.000000   
50%         0.001500       0.391300       0.909900       0.000000   
75%         0.002000       0.478300       0.967600       0.000000   
max         0.100000       1.000000       1.000000       1.000000   

                 734            735            736            737  \
count  190000.000000  190000.000000  190000.000000  190000.000000   
mean        0.704587       0.104526       0.006664       0.104526   
std         0.266625       0.305943       0.007502       0.305943   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.632300       0.000000       0.002800       0.000000   
50%         0.780100       0.000000       0.005500       0.000000   
75%         0.886600       0.000000       0.009200       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                 738  
count  190000.000000  
mean        0.012067  
std         0.010781  
min         0.000000  
25%         0.007100  
50%         0.010400  
75%         0.014200  
max         1.000000  

[8 rows x 739 columns]

## 1.2. Testing purpose data

Data upload for model testing purpose with cell execution time determination:

In [15]:
%%time
test_df = pd.read_csv('valami_test.txt.gz', sep='\t', header = None)

Wall time: 45.9 s


Taking a look what's inside:

In [8]:
test_df.head()

0       1    2    3    4    5    6    7    8    9     ...       729     730  \
0   ?  0.3571    0    0    1    0    0    0    0    0   ...    0.9336  0.0014   
1   ?  0.8571    0    0    0    1    0    0    0    0   ...    0.9414  0.0015   
2   ?  0.9286    0    0    1    0    0    0    0    0   ...    0.9492  0.0020   
3   ?  0.5714    0    0    0    1    0    0    0    0   ...    0.9336  0.0017   
4   ?  0.2143    0    0    1    0    0    0    0    0   ...    0.9180  0.0025   

      731     732  733     734  735     736  737     738  
0  0.4783  0.8378    0  0.6701    0  0.0046    0  0.0084  
1  0.3478  0.9784    0  0.9485    0  0.0074    0  0.0087  
2  0.4348  0.9964    0  0.0000    1  0.0000    1  0.0187  
3  0.2609  0.8919    0  0.7938    0  0.0101    0  0.0092  
4  0.8261  0.7928    0  0.5911    0  0.0064    0  0.0107  

[5 rows x 739 columns]

Obtaining data frame size:

In [9]:
test_df.shape

(95000, 739)

Data verification for columns using DataFrame information function: all columns (413+325=738) except classifier values are numeric (according to head() and shape functions information)

In [12]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95000 entries, 0 to 94999
Columns: 739 entries, 0 to 738
dtypes: float64(325), int64(413), object(1)
memory usage: 535.6+ MB


Some general information about columns values (maybe will help us later):

In [43]:
test_df.describe(include='all')

0             1             2             3             4    \
count   95000  95000.000000  95000.000000  95000.000000  95000.000000   
unique      1           NaN           NaN           NaN           NaN   
top         ?           NaN           NaN           NaN           NaN   
freq    95000           NaN           NaN           NaN           NaN   
mean      NaN      0.601348      0.427158      0.270295      0.179189   
std       NaN      0.244127      0.494668      0.444114      0.383513   
min       NaN      0.000000      0.000000      0.000000      0.000000   
25%       NaN      0.357100      0.000000      0.000000      0.000000   
50%       NaN      0.642900      0.000000      0.000000      0.000000   
75%       NaN      0.857100      1.000000      1.000000      0.000000   
max       NaN      1.000000      1.000000      1.000000      1.000000   

                 5             6             7             8             9    \
count   95000.000000  95000.000000  95000.000000  95000.000000  95000.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.086505      0.008053      0.020432      0.000200      0.003495   
std         0.281110      0.089375      0.141472      0.014141      0.059013   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%         0.000000      0.000000      0.000000      0.000000      0.000000   
50%         0.000000      0.000000      0.000000      0.000000      0.000000   
75%         0.000000      0.000000      0.000000      0.000000      0.000000   
max         1.000000      1.000000      1.000000      1.000000      1.000000   

            ...                729           730           731           732  \
count       ...       95000.000000  95000.000000  95000.000000  95000.000000   
unique      ...                NaN           NaN           NaN           NaN   
top         ...                NaN           NaN           NaN           NaN   
freq        ...                NaN           NaN           NaN           NaN   
mean        ...           0.894705      0.001748      0.375977      0.904226   
std         ...           0.080102      0.003780      0.173155      0.067406   
min         ...           0.238300      0.000000      0.000000      0.000000   
25%         ...           0.886700      0.001000      0.260900      0.841400   
50%         ...           0.925800      0.001500      0.391300      0.911700   
75%         ...           0.945300      0.002000      0.478300      0.967600   
max         ...           1.000000      1.000000      1.000000      1.000000   

                 733           734           735           736           737  \
count   95000.000000  95000.000000  95000.000000  95000.000000  95000.000000   
unique           NaN           NaN           NaN           NaN           NaN   
top              NaN           NaN           NaN           NaN           NaN   
freq             NaN           NaN           NaN           NaN           NaN   
mean        0.000021      0.704831      0.104916      0.006656      0.104916   
std         0.004588      0.266744      0.306446      0.007456      0.306446   
min         0.000000      0.000000      0.000000      0.000000      0.000000   
25%         0.000000      0.635700      0.000000      0.002800      0.000000   
50%         0.000000      0.780100      0.000000      0.005500      0.000000   
75%         0.000000      0.886600      0.000000      0.009200      0.000000   
max         1.000000      1.000000      1.000000      0.974300      1.000000   

                 738  
count   95000.000000  
unique           NaN  
top              NaN  
freq             NaN  
mean        0.012072  
std         0.009492  
min         0.000300  
25%         0.007100  
50%         0.010400  
75%         0.014200

# 2. Training data split to validation sets and fit it to model

Splitting model training data for parameters and classifier:

In [4]:
X = train_df.iloc[:, 1:]
y = train_df.iloc[:, :1]

Result shape verification - nothing was lost:

In [21]:
X.shape, y.shape

((190000, 738), (190000, 1))

Splitting model training data on the random base:
- the first sample - for model training with khown classifier value (70% of original population) 
- the second sample - for model testing (30% of original population)

Classifier values will be predicted for the second sample by model trained on the base of first sample.
Prediction will be compared with classifier data from the second sample

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state=17)

# 2.1. Logistic regression model

Loading logistic regression model (used when the response variable is categorical in nature):

In [24]:
logreg = LogisticRegression()

Model evaluation: quantifying the quality of predictions (average success rate of prediction by model on the base 'AUC ROC' scoring metric with cell execution time determination):

In [25]:
%%time
np.mean(cross_val_score(logreg, X_train, np.ravel(y_train), scoring='roc_auc', cv = 5))

Wall time: 3min 23s


0.6146669742091715

Model evaluation: quantifying the quality of predictions (maximum success rate of prediction by model on the base 'AUC ROC' scoring metric):

In [26]:
logreg.fit(X_train, np.ravel(y_train))
roc_auc_score(y_test, logreg.predict_proba(X_test)[:,1])

0.6323417054504198

##  2.2. "Decision tree" model

Loading "Decision tree" model with tree depth of 3 (the number of edges from the node to the tree's root node) and randomly generated (training multiple trees in an ensemble learner, where the features and samples are randomly sampled with replacement):

In [6]:
tree = DecisionTreeClassifier(max_depth = 3, random_state = 17)

Model evaluation: quantifying the quality of predictions (average success rate of prediction by model on the base 'AUC ROC' scoring metric with cell execution time determination):

In [7]:
%%time
np.mean(cross_val_score(tree, X_train, y_train, scoring='roc_auc', cv = 5))

Wall time: 43.7 s


0.5950253154124787

Model evaluation: quantifying the quality of predictions (maximum success rate of prediction by model on the base 'AUC ROC' scoring metric):

In [8]:
tree.fit(X_train, y_train)
roc_auc_score(y_test, tree.predict_proba(X_test)[:,1])

0.5958749132112721

Trying to improve quality of predictions by varying tree depth and number of parameters used for model fitting (features number variance led to sharp increase in calculation time):

In [9]:
%%time
tree_params = {'max_depth': np.arange(2,11)}
tree_grid = GridSearchCV(tree, tree_params, scoring='roc_auc', cv = 5, n_jobs=-1)
tree_grid.fit(X_train, y_train)

Wall time: 26min 29s


Optimized model parameters for better quality of predictions:

In [10]:
tree_grid.best_score_, tree_grid.best_params_

(0.6054853821258431, {'max_depth': 9})

New model success rate of prediction by model on the base 'AUC ROC' scoring metric:

In [11]:
roc_auc_score(y_test, tree_grid.predict_proba(X_test)[:,1])

0.602948120463012

## 2.3. k-nearest neighbors (kNN) model

Loading k-nearest neighbors (kNN) model:

In [24]:
kNN = KNeighborsClassifier(n_neighbors=10)

Model evaluation: quantifying the quality of predictions (average success rate of prediction by model on the base 'AUC ROC' scoring metric with cell execution time determination, cv is reduced to 2 because of performance):

In [ ]:
%%time
np.mean(cross_val_score(kNN, X_train, np.ravel(y_train), scoring='roc_auc', cv = 2))

Model evaluation: quantifying the quality of predictions (maximum success rate of prediction by model on the base 'AUC ROC' scoring metric):

In [ ]:
kNN.fit(X_train, np.ravel(y_train))
roc_auc_score(y_test, kNN.predict_proba(X_test)[:,1])

Trying to improve quality of predictions by varying number of neighbors used for model fitting (using all performance that we have):

In [ ]:
kNN_params = dict(n_neighbors=list(range(5, 21, 5)))
kNN_grid = GridSearchCV(kNN, kNN_params, cv=2, scoring='roc_auc', n_jobs = -1)
kNN_grid.fit(X_train, np.ravel(y_train))

Optimized model parameters for better quality of predictions:

In [ ]:
kNN_grid.best_score_, kNN_grid.best_params_

New model success rate of prediction by model on the base 'AUC ROC' scoring metric:

In [ ]:
roc_auc_score(y_test, kNN_grid.predict_proba(X_test)[:,1])

## 2.4. Random forest model

Loading "Random forest" model (differ from decision forest because it aggregates many decision trees to limit overfitting as well as error due to bias):

In [6]:
RFC = RandomForestClassifier(max_depth = 3, random_state = 17)

Model evaluation: quantifying the quality of predictions (average success rate of prediction by model on the base 'AUC ROC' scoring metric with cell execution time determination):

In [7]:
%%time
np.mean(cross_val_score(RFC, X_train, np.ravel(y_train), scoring='roc_auc', cv = 5))

Wall time: 27.5 s


0.6053606581959252

Trying to improve quality of predictions by varying tree depth and number of parameters used for model fitting (using all performance that we have):

In [8]:
%%time
RFC_params = {'max_depth': np.arange(2,11), 'max_features': np.arange(246,739,246)}
RFC_grid = GridSearchCV(RFC, RFC_params, scoring='roc_auc', cv = 5, n_jobs=-1)
RFC_grid.fit(X_train, np.ravel(y_train))

Wall time: 1h 32min 23s


Optimized model parameters for better quality of predictions:

In [9]:
RFC_grid.best_score_, RFC_grid.best_params_

(0.6369491328386508, {'max_depth': 10, 'max_features': 246})

New model success rate of prediction by model on the base 'AUC ROC' scoring metric:

In [10]:
roc_auc_score(y_test, RFC_grid.predict_proba(X_test)[:,1])

0.6366162757425133

Another predictions' quality improvement round by varying tree depth and number of parameters used for model fitting (using all performance that we have):

In [11]:
%%time
RFC_params = {'max_depth': np.arange(10,41, 5), 'max_features': np.arange(246,739,246)}
RFC_grid = GridSearchCV(RFC, RFC_params, scoring='roc_auc', cv = 5, n_jobs=-1)
RFC_grid.fit(X_train, np.ravel(y_train))

Wall time: 3h 19min 2s


In [12]:
RFC_grid.best_score_, RFC_grid.best_params_

(0.8035262212549641, {'max_depth': 40, 'max_features': 246})

New model success rate of prediction by model on the base 'AUC ROC' scoring metric:

In [13]:
roc_auc_score(y_test, RFC_grid.predict_proba(X_test)[:,1])

0.875427106305873

# 3. Classifier values prediction for model with maximum AUC ROC 

Random forest model was selected to finally predict classifier values:

In [16]:
predictions = RFC_grid.predict(test_df.iloc[:, 1:])

Prediction results were unloaded to csv file:

In [17]:
pd.DataFrame(predictions).to_csv('result.сsv', header = None, index = False)